tce/opset/base/implicit_lsu.cc

```diff
@@ -0,0 +1,300 @@
+/*
+    Copyright (c) 2002-2020 Tampere University.
+
+    This file is part of TTA-Based Codesign Environment (TCE).
+
+    Permission is hereby granted, free of charge, to any person obtaining a
+    copy of this software and associated documentation files (the "Software"),
+    to deal in the Software without restriction, including without limitation
+    the rights to use, copy, modify, merge, publish, distribute, sublicense,
+    and/or sell copies of the Software, and to permit persons to whom the
+    Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in
+    all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+ */
+/**
+ * @file implicit_lsu.cc
+ *
+ * Implicit LSU operation definitions.
+ *
+ * @author Kanishkan Vadivel (k.vadivel-no.spam-tue.nl)
+ * @author Barry de Bruin (e.d.bruin-no.spam-tue.nl)
+ */
+
+#include "OSAL.hh"
+#include "OperationGlobals.hh"
+
+#define RUNTIME_ERROR_WITH_INT(MESSAGE, DATA) {\
+       int len = strlen(MESSAGE) + 15;                \
+       char *tmpBuf = static_cast<char*>(alloca(len));\
+       snprintf(tmpBuf, len, "%s %d", MESSAGE, DATA); \
+       OperationGlobals::runtimeError(                \
+           tmpBuf, __FILE__, __LINE__, parent_);      \
+}
+
+#define RUNTIME_ERROR_WITH_LONG(MESSAGE, DATA) {\
+       int len = strlen(MESSAGE) + 21;                \
+       char *tmpBuf = static_cast<char*>(alloca(len));\
+       snprintf(tmpBuf, len, "%s %ld", MESSAGE, (long)DATA);    \
+       OperationGlobals::runtimeError(                \
+           tmpBuf, __FILE__, __LINE__, parent_);      \
+}
+
+DEFINE_STATE(IMPLICIT_LSU_STATE)
+    uint16_t loadStride;
+    uint16_t storeStride;
+    uint16_t loadAddr;
+    uint16_t storeAddr;
+INIT_STATE(IMPLICIT_LSU_STATE)
+    loadAddr = 0;
+    storeAddr = 0;
+    loadStride = 0;
+    storeStride = 0;
+END_INIT_STATE;
+
+END_DEFINE_STATE
+
+//////////////////////////////////////////////////////////////////////////////
+// set LSU configuration register
+//////////////////////////////////////////////////////////////////////////////
+OPERATION_WITH_STATE(ILD_ADDR, IMPLICIT_LSU_STATE)
+TRIGGER
+    STATE.loadAddr = UINT(1);
+END_TRIGGER;
+END_OPERATION_WITH_STATE(ILD_ADDR)
+
+OPERATION_WITH_STATE(ILD_STRIDE, IMPLICIT_LSU_STATE)
```

Comment:
Could we combine these to one operation to reduce needed opcodes to support the implicit load/stores? ILD_INIT or such?

Reply:
Moved to ILD_INIT and IST_INIT